orttraining/orttraining/core/optimizer/bias_dropout_fusion.cc

```diff
@@ -0,0 +1,178 @@
+
+    // matching for bias Add node
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "Add", {7}) ||
+        !graph_utils::IsSupportedProvider(node, GetCompatibleExecutionProviders()) ||
+        node.GetOutputEdgesCount() != 1) {
+      continue;
+    }
+
+    std::vector<NodeArg*> dropout_input, dropout_output;
+    const TensorShapeProto* input1_shape = node.MutableInputDefs()[0]->Shape();
+    const TensorShapeProto* input2_shape = node.MutableInputDefs()[1]->Shape();
+
+    if (input1_shape == nullptr ||
+        input2_shape == nullptr ||
+        input1_shape->dim_size() < 1 ||
+        input2_shape->dim_size() < 1) {
+      continue;
+    }
+
+    int last_dim_shape1 = input1_shape->dim_size() - 1;
+    int last_dim_shape2 = input2_shape->dim_size() - 1;
+    if (!utils::HasDimValue(input1_shape->dim(last_dim_shape1)) ||
+        !utils::HasDimValue(input2_shape->dim(last_dim_shape2)) ||
+        input1_shape->dim(last_dim_shape1).dim_value() != input2_shape->dim(last_dim_shape2).dim_value()) {
+      continue;
+    }
+
+    if (input1_shape->dim_size() == 1) {
+      dropout_input.push_back(node.MutableInputDefs()[1]);  // droput input
+      dropout_input.push_back(node.MutableInputDefs()[0]);  // bias
+    } else if (input2_shape->dim_size() == 1) {
+      dropout_input.push_back(node.MutableInputDefs()[0]);  // dropout input
+      dropout_input.push_back(node.MutableInputDefs()[1]);  // bias
+    } else {
+      continue;
+    }
+    Node& add_node = node;
+    nodes_to_fuse.push_back(add_node);
+
+    // matching for Dropout node
+    auto next_node_itr = node.OutputNodesBegin();
+    if (next_node_itr == node.OutputNodesEnd()) {
+      continue;
+    }
+
+    const Node& next_node = (*next_node_itr);
+    if (!(graph_utils::IsSupportedOptypeVersionAndDomain(next_node, "Dropout", {12}, kOnnxDomain) ||
+          graph_utils::IsSupportedOptypeVersionAndDomain(next_node, "TrainableDropout", {9}, kOnnxDomain)) ||
+        next_node.GetExecutionProviderType() != node.GetExecutionProviderType()) {
+      continue;
+    }
+
+    if (!graph.GetNodeOutputsInGraphOutputs(node).empty()) {
+      continue;
+    }
+
+    Node& dropout_node = const_cast<Node&>(next_node);
+    nodes_to_fuse.push_back(dropout_node);
+
+    dropout_output.push_back(dropout_node.MutableOutputDefs()[0]);
+    dropout_output.push_back(dropout_node.MutableOutputDefs()[1]);
+
+    bool is_onnx_dropout = dropout_node.OpType().compare("Dropout") == 0;
+
+    // matching for residual Add node
+    bool has_residual_add = false;
+    for (auto last_node_itr = dropout_node.OutputNodesBegin(); last_node_itr != dropout_node.OutputNodesEnd(); ++last_node_itr) {
+      const Node& last_node = (*last_node_itr);
+
+      if (graph_utils::IsSupportedOptypeVersionAndDomain(last_node, "Add", {7}) &&
+          last_node.GetExecutionProviderType() == node.GetExecutionProviderType()) {
+        const TensorShapeProto* input1_shape = last_node.InputDefs()[0]->Shape();
```

Comment:
>input1_shape [](start = 32, length = 12)

there is another variable in an enclosing scope with this name (same for input2_shape). could be confusing.